File: 3781-maximize-the-distance-between-points-on-a-square/maximize-the-distance-between-points-on-a-square.cpp
class Solution {
public:

    bool check(vector<long long>& arr, int side, int k, int limit){
        for (long long start:arr){
            long long end = start + side*4LL - limit;
            long long cur = start;

            for(int i=0; i<k-1;++i){
                auto it = ranges::lower_bound(arr, cur+limit);
                if (it == arr.end() || *it>end){
                    cur = -1;
                    break;
                }
                cur = *it;
            }
            if (cur>=0)
                return true;
        }
        return false;
    }
    int maxDistance(int side, vector<vector<int>>& points, int k) {
        vector<long long> arr;

        for(auto&p: points){
            int x = p[0], y=p[1];
            if (x==0){
                arr.push_back(y);
            }
            else if (y==side){
                arr.push_back(x + side);
            }
            else if (x==side){
                arr.push_back(side*3LL-y);
            }
            else{
                arr.push_back(side*4LL-x);
            }
        }
        sort(arr.begin(),arr.end());

        long long lo=0, hi = side;
        int ans = 0;

        while (lo<=hi){
            long long mid = (lo+hi)/2;
            if (check(arr,side,k,mid)){
                ans = mid;
                lo = mid+1;
            } 
            else{
                hi = mid-1;
            }
        }

        return ans;
        
        return -1;
    }
};

File: 3781-maximize-the-distance-between-points-on-a-square/maximize-the-distance-between-points-on-a-square.py
class Solution:
    def maxDistance(self, side: int, points: List[List[int]], k: int) -> int:

        arr = []
        
        for x,y in points:
            if (x==0): #L
                arr.append(y)
            elif (y==side): #U
                arr.append(side+x)
            elif (x==side): #R
                arr.append(3*side-y)
            else: #D
                arr.append(4*side-x)

        arr.sort()
 
        def check(limit):
            p = 4*side
            for start in arr:
                end = start + p - limit #(1)
                cur = start
                for _ in range(k-1):
                    idx = bisect_left(arr,cur+limit) #(2)
                    if (idx == len(arr)) or arr[idx]>end:
                        cur=-1
                        break
                    cur=arr[idx]
                if cur >=0:
                    return True
            return False

        lo, hi = 1 , side
        ans = 0

        while (lo<=hi):
            mid = (lo+hi)//2
            if (check(mid)==True):
                ans = mid
                lo = mid+1
            else:
                hi = mid-1
            
        return ans

"""
(1)
Ta cần: start - end >= limit, nhưng đây là mảng vòng tròn nên
start - (end-4p) >= limit
end - start <= 4p - limit
end <= 4p - limit + start

(2) bisect_left(arr,cur+limit)
Tìm vị trí trong mảng arr sao cho arr[vị trí đó]>=cur+limit
"""